### ISSUE TITLE
NestJs GRPC Kubernetes: Microservice stuck on Initialization when run inside kubernetes cluster

### ISSUE THREAD

**Issue by TMInnovations** [nestjs/nest#9726]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Current behavior

Starting the grpc microservice container with url: `0.0.0.0:5000` 

```
app.connectMicroservice<MicroserviceOptions>({
    transport: Transport.GRPC,
    options: {
      url: process.env.GRPC_ITEM_PACKAGE_URL,
      package: 'item',
      protoPath: join(__dirname, 'proto/item.proto'),
    },
  });
```

in a docker environment produces these logs:

```
[Nest] 24352  - 04.06.2022, 11:36:37     LOG [InstanceLoader] UserMessageModule dependencies initialized +3ms
[Nest] 24352  - 04.06.2022, 11:36:37     LOG [InstanceLoader] GraphQLModule dependencies initialized +3ms
[Nest] 24352  - 04.06.2022, 11:36:37     LOG [NestMicroservice] Nest microservice successfully started +132ms
[Nest] 24352  - 04.06.2022, 11:36:37     LOG [RoutesResolver] AppController {/}: +10ms
[Nest] 24352  - 04.06.2022, 11:36:37     LOG [RouterExplorer] Mapped {/health, GET} route +5ms
```
When running it inside a kubernetes cluster the logs seem to be stuck at line 2:

```
[Nest] 18 - 06/04/2022, 9:04:38 AM LOG [InstanceLoader] UserMessageModule dependencies initialized +1ms
[Nest] 18 - 06/04/2022, 9:04:38 AM LOG [InstanceLoader] GraphQLModule dependencies initialized +0ms
```

Any ideas why this would happen without an (timeout or similar) error message?

### Minimum reproduction code

-

### Steps to reproduce

Run any NestJs GRPC Microservice that provides a grpc package with url: `0.0.0.0:5000` inside a kubernetes cluster.

### Expected behavior

Microservice should be reachable on Port 5000 from other services. If there is a problem, an error log should be produced.

### Package

- [ ] I don't know. Or some 3rd-party package
- [ ] <code>@nestjs/common</code>
- [ ] <code>@nestjs/core</code>
- [X] <code>@nestjs/microservices</code>
- [ ] <code>@nestjs/platform-express</code>
- [ ] <code>@nestjs/platform-fastify</code>
- [ ] <code>@nestjs/platform-socket.io</code>
- [ ] <code>@nestjs/platform-ws</code>
- [ ] <code>@nestjs/testing</code>
- [ ] <code>@nestjs/websockets</code>
- [ ] Other (see below)

### Packages versions

```json
"@nestjs/microservices": "^8.4.4",
```

### Node.js version

16.14.2

### In which operating systems have you tested?

- [ ] macOS
- [X] Windows
- [X] Linux

**Comment by Tony133** [comment]
Hi @TMInnovations , for this type of question try to use the support [discord](https://discord.com/invite/G7Qnnhy) channel

The NestJS Core team uses GitHub to track bug reports, feature requests, and regressions.

**Comment by TMInnovations** [comment]
As I found out the issue now, I think this really is a bug.

I added a RabbitMQ Transporter to the application. However the url that I provided was wrong. Instead of throwing an error, the initialization process was just stuck. This is horrible to debug in production and happened to me today.

**Comment by micalevisk** [comment]
so to reproduce this we just need to supply any invalid url to `options.url`, regardless of the environment?

**Comment by Tony133** [comment]
@TMInnovations, if you can create a minimal reproduction of the issue in a clonable git repository, it would be of great help to the Nest Core Team.

**Comment by TMInnovations** [comment]
@micalevisk yes, tried it on local, docker and kubernetes environment. Initialization gets stuck and no error is thrown. really hard to debug if this happens for the first time + on your production server (faulty env var for broker-connection-url)

I will do my best to create a minrepo soon.

**Comment by delucca** [comment]
@TMInnovations as I've mentioned by @micalevisk your issue is related to #9749. Seems like when the connection fails within the transport we don't resolve the server promise and therefore the bootstrap of the application hangs.

I'll work on a solution for this 😄

**Comment by delucca** [comment]
@TMInnovations unfortunately I was not able to reproduce your issue 😢. Also, your issue is not related to #9749 as we expected, since after further investigation I was able to find out that that issue was only related to RabbitMQ itself.

Here is what I did to try to reproduce your problem:

- I've used a boilerplate app and added a gRPC microservice to it, without launching it (so, the gRPC server is unreachable), but after launching NestJS it works as expected (it doesn't hangs)
- I've tried to do the same, but pointing to an existing port that isn't a gRPC service (to evaluate a misconfiguration on your cluster), but NestJS throws an error as expected
- I've created a Minikube instance, deployed a sample gRPC server and them deployed that same app, to test it within a "Kubernetes-like" context, and it works (as you can see here):
![](https://i.imgur.com/hoqHR84.png)

Could you help me reproducing it? 😄
I think the best way, in this case, is for you to provide 2 Docker images + the manifests of them, so we can deploy into a Minikube to reproduce and debug the issue.

**Comment by TMInnovations** [comment]
@delucca as i wrote in one of my comments before my problem has nothing to do with grpc - just with rmq. Actually I had the exact same issue as #9749 and just was not sure where the problem came from (due to no logs or errors).
Thank you for investigating such a lot!

**Comment by delucca** [comment]
> @delucca as i wrote in one of my comments before my problem has nothing to do with grpc - just with rmq. Actually I had the exact same issue as #9749 and just was not sure where the problem came from (due to no logs or errors). Thank you for investigating such a lot!

Oh! Sorry. For some reason I understood that we also had an issue with gRPC 🤦🏻

Regarding RabbitMQ I've already solved that issue on #9751, we're just discussing the optimal way to handle the reconnection, but we're probably going to merge it soon 😄
